File: packages/plugin-spec/test/release.test.ts
import { createHash } from "node:crypto";
import { describe, expect, it } from "vitest";
import { parseReleaseManifest, verifyReference } from "../src/release.js";
import { contractRelease, kitRelease, pluginRelease, sidecarRelease, specRelease } from "./releaseFixture.js";

type Doc = Record<string, any>;
const URL = "https://github.com/soksak-ai/weather-plugin/releases/download/v0.0.1/plugin.json";

describe("flat component release documents", () => {
  it("uses one kind, id, and version shape for every component", () => {
    for (const [kind, raw] of [["plugin", pluginRelease()], ["sidecar", sidecarRelease()], ["kit", kitRelease()], ["contract", contractRelease()], ["spec", specRelease()]] as const) {
      expect(parseReleaseManifest(raw)).toMatchObject({ ok: true, value: { kind, version: "0.0.1" } });
    }
  });
  it("accepts exact runtime dependency references only for plugins and sidecars", () => {
    const dependency = { id: "weather-sidecar", version: "0.0.1", size: 123, sha256: "a".repeat(64) };
    expect(parseReleaseManifest(pluginRelease({ runtimeDependencies: { sidecars: [dependency] } }))).toMatchObject({ ok: true, value: { runtimeDependencies: { sidecars: [dependency] } } });
    expect(parseReleaseManifest(sidecarRelease({ runtimeDependencies: { sidecars: [dependency] } })).ok).toBe(true);
    expect(parseReleaseManifest(kitRelease({ runtimeDependencies: { sidecars: [dependency] } })).ok).toBe(false);
    expect(parseReleaseManifest(pluginRelease({ runtimeDependencies: { sidecars: [{ id: "weather-sidecar", version: "0.0.1" }] } })).ok).toBe(false);
    expect(parseReleaseManifest(pluginRelease({ runtimeDependencies: { sidecars: [{ ...dependency, url: URL }] } })).ok).toBe(false);
  });
  it("names manifest, artifacts, and evidence by bare filename inside the release directory", () => {
    expect(parseReleaseManifest(pluginRelease())).toMatchObject({ ok: true, value: {
      manifest: { file: "plugin.json", size: 12345 },
      artifacts: [{ target: "any", file: "weather-plugin-0.0.1.tgz", format: "tgz", manifest: "plugin.json" }],
      evidence: [{ file: "conformance-release.json" }],
    } });
    for (const file of ["", ".", "..", "sub/plugin.json", "../plugin.json", "/plugin.json", "plugin json", "plugin\\.json", URL]) {
      const artifact: Doc = pluginRelease(); artifact.artifacts[0].file = file;
      expect(parseReleaseManifest(artifact).ok, `artifact ${file}`).toBe(false);
      const evidence: Doc = pluginRelease(); evidence.evidence[0].file = file;
      expect(parseReleaseManifest(evidence).ok, `evidence ${file}`).toBe(false);
    }
  });
  it("binds source.repository to https://github.com/soksak-ai/<id>", () => {
    expect(parseReleaseManifest(pluginRelease())).toMatchObject({ ok: true, value: { source: { repository: "https://github.com/soksak-ai/weather-plugin" } } });
    for (const repository of [
      "https://github.com/example/weather-plugin",
      "https://github.com/soksak-ai/weather-sidecar",
      "https://github.com/soksak-ai/weather-plugin.git",
      "https://github.com/soksak-ai/weather-plugin/",
      "http://github.com/soksak-ai/weather-plugin",
      "https://github.com/Soksak-AI/weather-plugin",
    ]) {
      const result = parseReleaseManifest(pluginRelease({ source: { repository, commit: "a".repeat(40) } }));
      expect(result.ok, repository).toBe(false);
      if (!result.ok) expect(result.errors).toContain("release.source.repository: https://github.com/soksak-ai/weather-plugin required");
    }
  });
  it("verifies release bytes against a reference by size and sha256", () => {
    const bytes = new TextEncoder().encode(JSON.stringify(pluginRelease()));
    const sha256 = createHash("sha256").update(bytes).digest("hex");
    const reference = { id: "weather-plugin", version: "0.0.1", size: bytes.byteLength, sha256 };
    expect(() => verifyReference(bytes, reference)).not.toThrow();
    expect(() => verifyReference(bytes, { ...reference, size: bytes.byteLength + 1 })).toThrow("release size mismatch: weather-plugin@0.0.1");
    expect(() => verifyReference(bytes, { ...reference, sha256: "0".repeat(64) })).toThrow("release digest mismatch: weather-plugin@0.0.1");
  });
  it("rejects a url on any file reference", () => {
    for (const section of ["manifest", "artifacts", "evidence"]) {
      const withUrl: Doc = pluginRelease();
      const target = section === "manifest" ? withUrl.manifest : withUrl[section][0];
      target.url = URL;
      expect(parseReleaseManifest(withUrl).ok, `${section} url`).toBe(false);
      delete target.file;
      expect(parseReleaseManifest(withUrl).ok, `${section} url without file`).toBe(false);
    }
  });
  it("requires the manifest file to be the kind-specific manifest name", () => {
    for (const [raw, wrong] of [[pluginRelease(), "sidecar.json"], [sidecarRelease(), "plugin.json"], [kitRelease(), "manifest.json"]] as const) {
      const value: Doc = raw; value.manifest.file = wrong;
      expect(parseReleaseManifest(value).ok, wrong).toBe(false);
    }
  });
  it("requires unique sorted evidence files", () => {
    const duplicate: Doc = pluginRelease(); duplicate.evidence = [duplicate.evidence[0], { ...duplicate.evidence[0] }];
    expect(parseReleaseManifest(duplicate).ok).toBe(false);
    const unsorted: Doc = pluginRelease(); unsorted.evidence = [{ ...unsorted.evidence[0], file: "b.json" }, { ...unsorted.evidence[0], file: "a.json" }];
    expect(parseReleaseManifest(unsorted).ok).toBe(false);
    const sorted: Doc = pluginRelease(); sorted.evidence = [{ ...sorted.evidence[0], file: "a.json" }, { ...sorted.evidence[0], file: "b.json" }];
    expect(parseReleaseManifest(sorted).ok).toBe(true);
  });
  it("rejects nested identities, reports, ranges, and missing sizes", () => {
    expect(parseReleaseManifest({ ...pluginRelease(), plugin: { id: "weather-plugin", version: "0.0.1" } }).ok).toBe(false);
    const reports = pluginRelease(); reports.reports = reports.evidence; delete reports.evidence;
    expect(parseReleaseManifest(reports).ok).toBe(false);
    expect(parseReleaseManifest(pluginRelease({ version: "latest" })).ok).toBe(false);
    const missing = pluginRelease(); delete (missing.manifest as Record<string, unknown>).size;
    expect(parseReleaseManifest(missing).ok).toBe(false);
  });
});

File: go/platformspec/release_test.go
package platformspec

import (
	"encoding/json"
	"strings"
	"testing"
)

func releaseFixture(kind string) map[string]any {
	id := "soksak-" + kind + "-example"
	repository := "https://github.com/soksak-ai/" + id
	manifest := kind + ".json"
	target := "any"
	format := "tgz"
	if kind == "sidecar" {
		target = "aarch64-apple-darwin"
		format = "tar.gz"
	}
	return map[string]any{
		"kind": kind, "id": id, "version": "0.0.1",
		"manifest":  map[string]any{"file": manifest, "size": 1, "sha256": strings.Repeat("a", 64)},
		"source":    map[string]any{"repository": repository, "commit": strings.Repeat("b", 40)},
		"artifacts": []any{map[string]any{"target": target, "file": "archive.tar.gz", "size": 2, "sha256": strings.Repeat("c", 64), "format": format, "manifest": manifest}},
		"evidence":  []any{map[string]any{"file": "report.json", "size": 3, "sha256": strings.Repeat("d", 64)}},
	}
}
func encoded(t *testing.T, value any) []byte {
	t.Helper()
	body, err := json.Marshal(value)
	if err != nil {
		t.Fatal(err)
	}
	return body
}
func rejects(t *testing.T, name string, value map[string]any) {
	t.Helper()
	if _, err := ParseReleaseManifest(encoded(t, value)); err == nil {
		t.Fatalf("%s accepted", name)
	}
}
func TestReleaseManifestUsesTheFlatShapeForEveryKind(t *testing.T) {
	for _, kind := range []string{"plugin", "sidecar", "kit", "contract", "spec"} {
		if _, err := ParseReleaseManifest(encoded(t, releaseFixture(kind))); err != nil {
			t.Fatalf("%s: %v", kind, err)
		}
	}
}
func TestReleaseManifestRejectsNestedIdentityReportsAndWrongTargets(t *testing.T) {
	nested := releaseFixture("plugin")
	nested["plugin"] = map[string]any{"id": "old"}
	rejects(t, "nested identity", nested)
	reports := releaseFixture("plugin")
	reports["reports"] = reports["evidence"]
	delete(reports, "evidence")
	rejects(t, "reports", reports)
	wrong := releaseFixture("plugin")
	wrong["artifacts"].([]any)[0].(map[string]any)["target"] = "aarch64-apple-darwin"
	rejects(t, "native plugin target", wrong)
}
func TestReleaseManifestBindsSourceRepositoryToTheOrgAndID(t *testing.T) {
	for name, repository := range map[string]string{
		"fork owner":       "https://github.com/example/soksak-plugin-example",
		"id mismatch":      "https://github.com/soksak-ai/soksak-sidecar-example",
		"git suffix":       "https://github.com/soksak-ai/soksak-plugin-example.git",
		"trailing slash":   "https://github.com/soksak-ai/soksak-plugin-example/",
		"plain http":       "http://github.com/soksak-ai/soksak-plugin-example",
		"upper-case owner": "https://github.com/Soksak-AI/soksak-plugin-example",
	} {
		value := releaseFixture("plugin")
		value["source"] = map[string]any{"repository": repository, "commit": strings.Repeat("b", 40)}
		rejects(t, name, value)
	}
}
func TestReleaseManifestFileNamesFollowOneGrammar(t *testing.T) {
	for _, file := range []string{".", "..", "a b", "x?y", "x#y", "r\u00e9port.json", "%2e%2e", "report/../x"} {
		value := releaseFixture("plugin")
		value["evidence"] = []any{map[string]any{"file": file, "size": 3, "sha256": strings.Repeat("d", 64)}}
		rejects(t, "file "+file, value)
	}
	for _, file := range []string{"a", "...", ".hidden", "Report_1.tar.gz", "a-b_c.d"} {
		value := releaseFixture("plugin")
		value["evidence"] = []any{map[string]any{"file": file, "size": 3, "sha256": strings.Repeat("d", 64)}}
		if _, err := ParseReleaseManifest(encoded(t, value)); err != nil {
			t.Fatalf("file %s: %v", file, err)
		}
	}
}
func TestReleaseManifestRejectsURLAndNonBareFileNames(t *testing.T) {
	withURL := releaseFixture("plugin")
	withURL["evidence"] = []any{map[string]any{"url": "https://github.com/soksak-ai/x/releases/download/v0.0.1/report.json", "file": "report.json", "size": 3, "sha256": strings.Repeat("d", 64)}}
	rejects(t, "url on integrity reference", withURL)
	slash := releaseFixture("plugin")
	slash["evidence"] = []any{map[string]any{"file": "reports/report.json", "size": 3, "sha256": strings.Repeat("d", 64)}}
	rejects(t, "file with slash", slash)
	empty := releaseFixture("plugin")
	empty["evidence"] = []any{map[string]any{"file": "", "size": 3, "sha256": strings.Repeat("d", 64)}}
	rejects(t, "empty file", empty)
	digest := releaseFixture("plugin")
	digest["evidence"] = []any{map[string]any{"file": "report.json", "size": 3, "sha256": "abc"}}
	rejects(t, "short digest", digest)
	size := releaseFixture("plugin")
	size["evidence"] = []any{map[string]any{"file": "report.json", "size": 0, "sha256": strings.Repeat("d", 64)}}
	rejects(t, "zero size", size)
}
func TestReleaseManifestRequiresTheKindManifestFile(t *testing.T) {
	other := releaseFixture("plugin")
	other["manifest"] = map[string]any{"file": "sidecar.json", "size": 1, "sha256": strings.Repeat("a", 64)}
	rejects(t, "manifest file of another kind", other)
	artifact := releaseFixture("plugin")
	artifact["artifacts"].([]any)[0].(map[string]any)["manifest"] = "sidecar.json"
	rejects(t, "artifact manifest of another kind", artifact)
}
func TestReleaseManifestRuntimeDependenciesAreReleaseReferences(t *testing.T) {
	reference := func(id string) map[string]any {
		return map[string]any{"id": id, "version": "0.0.1", "size": 4, "sha256": strings.Repeat("e", 64)}
	}
	ok := releaseFixture("plugin")
	ok["runtimeDependencies"] = map[string]any{"plugins": []any{reference("a-plugin"), reference("b-plugin")}, "sidecars": []any{reference("a-sidecar")}}
	if _, err := ParseReleaseManifest(encoded(t, ok)); err != nil {
		t.Fatal(err)
	}
	withURL := releaseFixture("plugin")
	url := reference("a-plugin")
	url["url"] = "https://github.com/soksak-ai/a-plugin/releases/download/v0.0.1/release.json"
	withURL["runtimeDependencies"] = map[string]any{"plugins": []any{url}}
	rejects(t, "url on release reference", withURL)
	unsorted := releaseFixture("plugin")
	unsorted["runtimeDependencies"] = map[string]any{"plugins": []any{reference("b-plugin"), reference("a-plugin")}}
	rejects(t, "unsorted release references", unsorted)
	kit := releaseFixture("kit")
	kit["runtimeDependencies"] = map[string]any{"plugins": []any{reference("a-plugin")}}
	rejects(t, "runtime dependencies on kit", kit)
}
